server/controllers/caseNotesController.ts

```diff
@@ -20,63 +29,332 @@ export default class CaseNotesController {
-    const queryParams: PagedListQueryParams = {}
-    const { clientToken } = res.locals
-    const userToken = res.locals.user.token
-
-    if (req.query.page) queryParams.page = +req.query.page
-    if (req.query.sort) queryParams.sort = req.query.sort as string
-    if (req.query.type) queryParams.type = req.query.type as string
-    if (req.query.subType) queryParams.subType = req.query.subType as string
-    if (req.query.startDate) queryParams.startDate = req.query.startDate as string
-    if (req.query.endDate) queryParams.endDate = req.query.endDate as string
-    if (req.query.showAll) queryParams.showAll = Boolean(req.query.showAll)
-
-    // Get prisoner data for banner and for use in alerts generation
-    const prisonerData = await this.prisonerSearchService.getPrisonerDetails(clientToken, req.params.prisonerNumber)
-
-    // Set role based permissions
-    const canDeleteSensitiveCaseNotes = userHasRoles([Role.DeleteSensitiveCaseNotes], res.locals.user.userRoles)
-
-    // If user cannot view this prisoner's case notes, redirect to 404 page
-    if (!canViewCaseNotes(res.locals.user, prisonerData)) {
-      return res.render('notFound.njk', {
-        url: req.headers.referer || `/prisoner/${prisonerData.prisonerNumber}`,
+  public displayCaseNotes(): RequestHandler {
+    return async (req: Request, res: Response) => {
+      // Parse query params for paging, sorting and filtering data
+      const queryParams: PagedListQueryParams = {}
+      const { clientToken } = res.locals
+      const userToken = res.locals.user.token
+
+      if (req.query.page) queryParams.page = +req.query.page
+      if (req.query.sort) queryParams.sort = req.query.sort as string
+      if (req.query.type) queryParams.type = req.query.type as string
+      if (req.query.subType) queryParams.subType = req.query.subType as string
+      if (req.query.startDate) queryParams.startDate = req.query.startDate as string
+      if (req.query.endDate) queryParams.endDate = req.query.endDate as string
+      if (req.query.showAll) queryParams.showAll = Boolean(req.query.showAll)
+
+      // Get prisoner data for banner and for use in alerts generation
+      const prisonerData = await this.prisonerSearchService.getPrisonerDetails(clientToken, req.params.prisonerNumber)
+
+      // Set role based permissions
+      const canDeleteSensitiveCaseNotes = userHasRoles([Role.DeleteSensitiveCaseNotes], res.locals.user.userRoles)
+
+      // If user cannot view this prisoner's case notes, redirect to 404 page
+      if (!canViewCaseNotes(res.locals.user, prisonerData)) {
+        return res.render('notFound.njk', {
+          url: req.headers.referer || `/prisoner/${prisonerData.prisonerNumber}`,
+        })
+      }
+
+      let addCaseNoteLinkUrl: string
+      if (canAddCaseNotes(res.locals.user, prisonerData)) {
+        addCaseNoteLinkUrl = `/prisoner/${prisonerData.prisonerNumber}/add-case-note`
+      }
```

Comment:
```suggestion
      const addCaseNoteLinkUrl = canAddCaseNotes(res.locals.user, prisonerData) ? `/prisoner/${prisonerData.prisonerNumber}/add-case-note` : undefined
```

I get nervous when variables are set with let and then mutated, makes me worried about where else they are changed